core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java

```diff
@@ -0,0 +1,626 @@
+   */
+  public RelDataType getWiderTypeFor(List<RelDataType> typeList, boolean stringPromotion) {
+    assert typeList.size() > 1;
+    RelDataType resultType = typeList.get(0);
+
+    List<RelDataType> target = stringPromotion ? partitionByCharacter(typeList) : typeList;
+    for (RelDataType tp : target) {
+      resultType = getWiderTypeForTwo(tp, resultType, stringPromotion);
+      if (null == resultType) {
+        return null;
+      }
+    }
+    return resultType;
+  }
+
+  private List<RelDataType> partitionByCharacter(List<RelDataType> types) {
+    List<RelDataType> withCharacterTypes = Lists.newArrayList();
+    List<RelDataType> nonCharacterTypes = Lists.newArrayList();
+
+    for (RelDataType tp : types) {
+      if (SqlTypeUtil.hasCharacter(tp)) {
+        withCharacterTypes.add(tp);
+      } else {
+        nonCharacterTypes.add(tp);
+      }
+    }
+
+    List<RelDataType> partitioned = Lists.newArrayList();
+    partitioned.addAll(withCharacterTypes);
+    partitioned.addAll(nonCharacterTypes);
+    return partitioned;
+  }
+
+  /**
+   * Check if the types and families can have implicit type cast. We will check the type one
+   * by one, that means the 1th type and 1th family, 2th type and 2th family, and so on.
+   *
+   * @param types    data type need to check.
+   * @param families desired type families list.
+   * @return if every pair of (type, family) can do type coercion.
+   */
+  boolean canImplicitTypeCast(List<RelDataType> types, List<SqlTypeFamily> families) {
+    boolean needed = false;
+    if (types.size() != families.size()) {
+      return false;
+    }
+    for (Pair<RelDataType, SqlTypeFamily> pair : Pair.zip(types, families)) {
+      RelDataType implicitType = implicitCast(pair.left, pair.right);
+      if (null == implicitType) {
+        return false;
+      }
+      needed = pair.left != implicitType || needed;
+    }
+    return needed;
+  }
+
+  /**
+   * Type coercion based on the inferred type from passed in operand and the {@link SqlTypeFamily}
+   * defined in the checkers, i.e. the {@link org.apache.calcite.sql.type.FamilyOperandTypeChecker}.
+   * Caution that We do not cast from numeric and numeric.
+   * See <a href="https://docs.google.com/spreadsheets/d/1GhleX5h5W8-kJKh7NMJ4vtoE78pwfaZRJl88ULX_MgU/edit?usp=sharing">
+   * CalciteImplicitCasts</a> for the details.
+   *
+   * @param in       inferred operand type.
+   * @param expected expected {@link SqlTypeFamily} of registered FUNC.
+   * @return common type of implicit cast, null if we do not find one.
+   */
+  public RelDataType implicitCast(RelDataType in, SqlTypeFamily expected) {
+    List<SqlTypeFamily> numericFamilies = Lists.newArrayList(
+        SqlTypeFamily.NUMERIC,
+        SqlTypeFamily.DECIMAL,
+        SqlTypeFamily.APPROXIMATE_NUMERIC,
+        SqlTypeFamily.EXACT_NUMERIC,
+        SqlTypeFamily.INTEGER);
+    List<SqlTypeFamily> dateTimeFamilies = Lists.newArrayList(SqlTypeFamily.DATE,
```

Comment:
Can boolean be cast to decimal?  following this logic, i can deduce that boolean can cast to decimal.